### DISCUSSION TITLE
Quantized Models

### DISCUSSION THREAD

**Discussion by kroggen** [VsonicV/es-fine-tuning-paper#15] [General]
A place to discuss the fine-tuning of quantized models

We can divide them into 2 categories:

* Discrete (Q8, Q5, Q4...)
* Continuous (FP8, FP4, NVFP4...)

**Comment by kroggen** [comment]
Here is a similar work made with NVFP4:

https://x.com/yukangchen_/status/1977934836737122643

**Comment by VsonicV** [comment]
@kroggen Thanks for creating this thread. We were also planning to try ES on quantized models (discrete ones), which were usually infeasible for backprop based methods. This is a very important and promising use case.

**Reply by kroggen** [reply]
It is exactly what I am doing

Using a model in GGUF format with Q8 quantization and inference via `llama-cpp-python`

The minimal noise is 1 unit for int8  (-1, 0, 1) adjusted to modify only a fraction of the parameters

But it turns out to be super slow, 40 minutes for each seed! (not each iteration) on a RTX 3090

**Reply by kroggen** [reply]
And the reward was 0.0 when modifying 10% of the parameters. It must target less parameters, like 1% or even less. And thus it may require way more seeds to find the one(s) that find the exact parameters that need to be modified AND to the right direction (+1 or -1)

**Reply by kroggen** [reply]
That approach I posted above (from NVIDIA) may be better. They use a combination of backprop AND noise. Very interesting

**Reply by kroggen** [reply]
There is also this new alternative technique to get gradients for quantized models:

https://x.com/ShamKakade6/status/1978531483909259302

**Reply by VsonicV** [reply]
@kroggen Wow, everything you shared here is very valuable for people who are interested in working on this topic. Thanks for sharing this with the community! Regarding the current issues you encountered, the change ratio (with regards to the original weight magnitude) of the minimum change unit in quantized model is significantly larger than those in our current continuous implementations. This makes the perturbation step too big. Maybe as you suggested here, if we cannot do too much in reducing the noise scale, we can do something about the noise sparsity, i.e., for each block (we can define it) in the original architecture, we only perturb one weight.
